### DISCUSSION TITLE
ASCIIMath plugin and markdown delimiter/format discussion

### DISCUSSION THREAD

**Discussion by meghprkh** [remarkjs/remark#572] [Ideas] [answered]
Hi!

I am developing a remark/rehype plugin for [ASCIIMath](https://asciimath.org) and was wondering if there is any standard markdown way and what would be your recommendations for it.

Context: `remark-math` uses `$` (eg `$ x^2 $`) for inline Latex and `$$` for Latex blocks

ASCIIMath is very convenient for quickly typing maths in and is spiritually equivalent to markdown for Math.

Like the requirements for Markdown+Latex, most Markdown+ASCIIMath documents will involve considerable amount of inlined math and need a convenient way to use it

Now I will describe all the options I could research/come up with

## Option 1: Use a special delimiter

[Notable.app](https://cheatsheet.md/notable.pdf) uses `&` and `&&` instead of `$` and `$$` as delimiters to pass to ASCIIMath.

This is extremely convenient but takes up a new delimiter.

## Option 2: Support this using ``$`<...>`$``

I came up with this custom format, since ASCIIMath originally recommends using `` ` `` as the separator

If the latex block starts and ends with back-ticks, we pass it through ASCIIMath

```
$`x/y`$ 
$$`
x / y
`$$
```

Since latex does not care about spaces, if we put a space between `$` and `` ` ``, it should not go through ASCIIMath. This should also ideally not be breaking as LaTeX does not use backticks I think.

I have also been able to implement this extremely easily using a rehype plugin.

## Option 3: Pass all math through ASCIIMath

[MarkdownApp](https://marknoteapp.com/doc/help) seems to pass all latex through ASCIIMath

This probably works because ASCIIMath is in a way a superset of Latex and should be keeping latex commands as it is, but it should break as explained below

**NOTE**: I couldn't get ASCIIMath working on web site

## Option 4: Like option 3, but allow escaping certain Latex

Pass everything through ASCIIMath, except formulas escaped using backticks

This has the advantage of using the same delimiter and allowing escapes directly to LaTeX

This becomes particularly convenient because some formulas otherwise need to be escaped to be used as ASCIIMath. Example

`\frac{1 / 2}{3}`

With ASCIIMath this becomes:
![image](https://user-images.githubusercontent.com/5979605/99938446-ce513800-2d8d-11eb-9158-2ed11ee50021.png)

Without ASCIIMath it stays as intended:
![image](https://user-images.githubusercontent.com/5979605/99938480-e0cb7180-2d8d-11eb-9526-bb5c4276bd67.png)

Thus, this is convenient for new writers, but is potentially breaking for people who are already using Markdown+Latex

---

I would like your suggestions on this :smile:

**Comment by ChristianMurphy** [comment, marked as answer]
@meghprkh it sounds like you may be able to use `remark-math`?
Your example ``$\`x/y\`$`` for example already can be parsed by `remark-math`.
https://astexplorer.net/#/gist/cfadde5ebebb56064f0b5361c63971b5/25647671206d6f1aef205121c2b91bef8f5cf520

Once it's parsed, you can do any additional preprocessing needed, and pass the values to asciimath.

**Reply by meghprkh** [reply]
Yes that is my plan. It also requires a very small code snippet. However I was wondering if we could settle on a delimiter for the same and use that consistently. An/or commit the code-snippet as a subpackage to `remark-math`.

**Reply by ChristianMurphy** [reply]
> Yes that is my plan
> if we could settle on a delimiter for the same and use that consistently

I'm not sure I follow, for `remark-math`, `$` is the delimiter, and it sounds like it works for your use case.
What needs to be settled?

`remark-math` the parser doesn't specify what language is inside the math blocks, that is decided by other plugins, currently LaTeX based languages have plugins, but can also be asciimath or any other text based math language.

> An/or commit the code-snippet as a subpackage to `remark-math`

I'm open to the idea of having it as a subpackage in the remark-math repo.
@remarkjs/core thoughts? :thought_balloon:

**Reply by meghprkh** [reply]
> `remark-math` the parser doesn't specify what language is inside the math blocks, that is decided by other plugins, currently LaTeX based languages have plugins, but can also be asciimath or any other text based math language.

What I was intending to say is people using ASCIIMath would also likely be using latex. So, I wished to get an opinion on if it will make sense to use different delimiters other than `$` like option 1 using `&` or shall be use `$` with an escape like I suggested.

> An/or commit the code-snippet as a subpackage to remark-math

I will also create a PR there with the `$` plus escape (`` ` ``) mechanism :+1: Thanks :smile:
